sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,549 @@
+                          uint8_t nstreams, uint32_t reliable, bool remap) {
+  usb_logging_ctx_t *ctx = &the_ctx;
+
+  // Validate input arguments.
+  TRY_CHECK(nstreams > 0u && nstreams <= kUSBLogMaxStreams);
+
+  // Set up the software layers, if required.
+  if (usbutils) {
+    // usb_testutils layer already initialized, so an endpoint number must be
+    // supplied.
+    ctx->usbutils = usbutils;
+    TRY_CHECK(ep_first >= 1u && USBDEV_NUM_ENDPOINTS >= ep_first + nstreams);
+  } else {
+    // Pointers to context for lower software layers.
+    ctx->usbutils = &usbdev;
+    ep_first = 1u;
+
+    TRY(dif_pinmux_init(
+        mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+    pinmux_testutils_init(&pinmux);
+    TRY(dif_pinmux_input_select(&pinmux,
+                                kTopEarlgreyPinmuxPeripheralInUsbdevSense,
+                                kTopEarlgreyPinmuxInselIoc7));
+
+    // Total length of the configuration descriptor; validate caller buffer.
+    size_t cfg_len = USB_CFG_DSCR_LEN +
+                     nstreams * (USB_INTERFACE_DSCR_LEN + USB_EP_DSCR_LEN);
+    TRY_CHECK(cfg_len <= sizeof(config_descriptors));
+
+    // Configuration Descriptor header.
+    uint8_t *cfg = config_descriptors;
+    uint8_t head[USB_CFG_DSCR_LEN] = {
+        USB_CFG_DSCR_HEAD((uint16_t)cfg_len, (uint8_t)nstreams)};
+    memcpy(cfg, head, USB_CFG_DSCR_LEN);
+    cfg += USB_CFG_DSCR_LEN;
+
+    // Followed by programmatically-generated list of interface descriptors.
+    for (uint8_t id = 0U; id < nstreams; ++id) {
+      uint8_t ep_in = ep_first + id;
+      // Description of a single interface.
+      uint8_t int_dscr[USB_INTERFACE_DSCR_LEN + USB_EP_DSCR_LEN] = {
+          VEND_INTERFACE_DSCR(id, 1, 0x50, 1),
+          // Declare only an IN endpoint because we don't require OUT traffic
+          // and any character echo would just be wasting bus bandwidth.
+          USB_EP_DSCR(1, ep_in, (uint8_t)kUsbTransferTypeBulk,
+                      USBDEV_MAX_PACKET_SIZE, 0),
+      };
+
+      // Append interface descriptor to the configuration descriptor.
+      memcpy(cfg, int_dscr, sizeof(int_dscr));
+      cfg += sizeof(int_dscr);
+    }
+
+    // Call `usbdev_init` here so that DPI will not start until the
+    // simulation has finished all of the printing, which takes a while
+    // if `--trace` was passed in.
+    TRY(usb_testutils_init(ctx->usbutils, /*pinflip=*/false,
+                           /*en_diff_rcvr=*/true,
+                           /*tx_use_d_se0=*/false));
+    TRY(usb_testutils_controlep_init(&usbdev_control, ctx->usbutils, 0,
+                                     config_descriptors, cfg_len, NULL,
+                                     0u));  // No need for test descriptor.
+  }
+
+  // Initialize list of free buffers.
+  free_list_init(ctx);
+
+  // Initialize the time of the next flush on each stream; stagger them to
+  // even out bus usage.
```

Comment:
This staggering is for the initial state. Do you have any mechanism to even out the bus usage in the steady state? If so, why do we need this up front? If not, isn't this initial staggering going to be irrelevant once the steady state is reached?

Reply:
I've changed the comment to make this clearer, hopefully. It's only a small point about making the traffic less bursty and reducing contention among streams. The timers run independently but they will drift relative to each other, and the nature of the logging traffic on individual streams is unknown; there probably won't be a steady state. Staggering them at the start is just an inexpensive effort to lend a helping hand.